python/cudf/cudf/core/mixins/mixin_factory.py

```diff
@@ -0,0 +1,163 @@
+        name will be used to define the following attributes as shown in the
+        example below:
+            - f'_{category_name}_DOCSTRINGS'
+            - f'_VALID_{category_name}S'
+            - f'_SUPPORTED_{category_name}S'
+    operation_name : str
+        The name given to the core function implementing this category of
+        operations.  The corresponding function is the entrypoint for child
+        classes.
+    supported_ops : List[str]
+        The list of valid operations that subclasses of the resulting mixin may
+        request to be implemented.
+
+    Examples
+    --------
+    >>> MyFoo = _create_delegating_mixin(
+    ...     "MyFoo", "MyFoo's docstring", "FOO", "do_foo", {"foo1", "foo2"}
+    ... )
+
+    >>> # The above is equivalent to defining a class like so:
+    ... class MyFoo:
+    ...     # The set of valid foo operations.
+    ...     _VALID_FOOS = {"foo1", "foo2"}
+    ...
+    ...     # This is the method for child classes to override. Note that the
+    ...     # first parameter is always called "op".
+    ...     def do_foo(self, op, *args, **kwargs):
+    ...         raise NotImplementedError
+
+    >>> # MyFoo can be used as follows.
+    >>> class BarImplementsFoo(MyFoo):
+    ...     _SUPPORTED_FOOS = ("foo1",)
+    ...     _FOO_DOCSTRINGS = {"ret": "42"}
+    ...
+    ...     # This method's docstring will be formatted and used for all valid
+    ...     # operations. The `op` key is always automatically filled in, while
+    ...     # other keys are formatted using _FOO_DOCSTRINGS.
+    ...     def do_foo(self, op, *args, **kwargs):
+    ...         '''Perform the operation {op}, which returns {ret}.'''
+    ...         return 42
+
+    >>> bar = BarImplementsFoo()
+    >>> print(bar.foo1())
+    42
+
+    >>> # This will raise an AttributeError because foo2 is not supported by
+    >>> # the BarImplementsFoo subclass of MyFoo.
+    >>> # print(bar.foo2())
+
+    >>> # The docstring is formatted with the operation name as well as any
+    >>> # additional keys provided via the _FOO_DOCSTRINGS parameter.
+    >>> print(bar.foo1.__doc__)
+    Perform the operation foo1, which returns 42.
+    """
+
+    docstring_attr = f"{category_name}_DOCSTRINGS"
+    validity_attr = f"_VALID_{category_name}S"
+
+    class OperationMixin:
+        @classmethod
+        def _add_operation(cls, operation):
+            # This function creates operations on-the-fly.
+
+            # Generate a signature without the `op` parameter.
+            base_operation = getattr(cls, operation_name)
+            base_params = inspect.signature(base_operation).parameters.values()
+            params = [p for p in base_params if p.name != "op"]
+            signature = inspect.Signature(parameters=params)
+
+            # Generate the list of arguments forwarded to the base operation.
+            arglist = ", ".join(
+                [
+                    f"{key}={key}"
```

Comment:
No, if you look below you'll see that `arglist` is formatted into the string containing the defined function so that its call to the `base_operation` passes all arguments up by keyword. For example, here's what you want this to look like
```
def max(self, arg1=defaultarg1, arg2=defaultarg2, *args, **kwargs):
    return self._reduce(op='max', arg1=arg1, arg2=arg2)
```
The `arglist` corresponds to the `arg1=arg1, arg2=arg2`. We don't want to set the default parameters here because then they would override arguments provided by the user, we actually want to forward along the provided arguments.

Reply:
Ooh, ok. Thanks for explaining!